### PR TITLE
[terra-framework] Enhance Theme Variables and Styles

### PR DESCRIPTION
**Pull request by BenBoersma**
### Major Changes
* Rename, remove, or add theme variables
* Add Upgrade guides that highlight theme variable changes. 
* Minute render changes that shift certain DOM elements.
* New CSS Classes for missing styles

#### Note
I did not make all of these changes. To help with discussion, anyone that contributed to theme updates is an assignee of this PR.

Thanks for contributing to Terra. 
@cerner/terra

[CONTRIBUTORS.md]: ../blob/master/CONTRIBUTORS.md
[License]: ../blob/master/LICENSE

### CODE REVIEW

File: packages/terra-time-input/src/TimeInput.module.scss [hunk 1]
```diff
@@ -64,66 +92,77 @@
...
+    caret-color: var(--terra-time-input-caret-color, #1c1f21);
     color: var(--terra-time-input-color, #1c1f21);
     display: table;
-    height: var(--terra-time-input-mobile-height, 1.857rem);
-    min-width: var(--terra-time-input-width-input, 3.57em);
-    padding-bottom: var(--terra-time-input-input-padding-bottom, 0.357em);
-    padding-left: var(--terra-time-input-input-padding-left, 0.357em);
-    padding-right: var(--terra-time-input-input-padding-right, 0.357em);
-    padding-top: var(--terra-time-input-input-padding-top, 0.357em);
+    height: var(--terra-time-input-mobile-height, 2.1428571429rem);
```
Comment (bjankord) [inline packages/terra-time-input/src/TimeInput.module.scss]: Let's remove the word, `mobile` from these variables

Comment (bjankord) [reply packages/terra-time-input/src/TimeInput.module.scss]: If the variable without the `mobile` word already exists in this file, like `--terra-time-input-line-height`, let's update it to `--terra-time-input-desktop-line-height`.

File: packages/terra-layout/src/LayoutSlidePanel.module.scss [hunk 2]
```diff
@@ -39,28 +39,49 @@
...
+      box-shadow: var(--terra-layout-slide-panel-overlay-panel-rtl-box-shadow, -1px 0 0 0 #1c1f21, 0 0 5px 1px #000, 0 0 12px 2px rgba(0, 0, 0, 0.8));
+    }
+    /* stylelint-enable selector-max-compound-selectors */
   }
 
   // Squish
   .is-open.is-squish {
     > .content {
-      width: calc(100% - #{var(--layout-slide-panel-desktop-panel-width, 18.571rem)} - 0.07143rem);
+      width: calc(100% - #{var(--terra-layout-slide-panel-panel-width, 18.571rem)} - #{var(--terra-layout-slide-panel-squish-panel-border-right, 1px)});
```
Comment (bjankord) [inline packages/terra-layout/src/LayoutSlidePanel.module.scss]: I think this should be updated to `--terra-layout-slide-panel-squish-panel-border-right-width`.  If we leave the width off, people may think they can set this value to `2px solid #f00` where we just want the width value to use in the calculation.

File: packages/terra-layout/src/LayoutSlidePanel.module.scss [hunk 3]
```diff
@@ -39,28 +39,49 @@
...
 
     > .panel {
-      border-right: 0.07143rem solid var(--layout-slide-panel-panel-border-color, #d3d4d5);
+      background-color: var(--terra-layout-slide-panel-squish-panel-background-color, #fff);
+      border-right-color: var(--terra-layout-slide-panel-squish-panel-border-right-color, #d3d4d5);
+      border-right-style: solid;
+      border-right-width: var(--terra-layout-slide-panel-squish-panel-border-right-width, 1px);
       transition: none;
-      width: calc(0.07143rem + #{var(--layout-slide-panel-desktop-panel-width, 18.571rem)});
+      width: calc(#{var(--terra-layout-slide-panel-squish-panel-border-right, 1px)} + #{var(--terra-layout-slide-panel-panel-width, 18.571rem)});
```
Comment (bjankord) [inline packages/terra-layout/src/LayoutSlidePanel.module.scss]: I think this should be updated to `--terra-layout-slide-panel-squish-panel-border-right-width`.  If we leave the width off, people may think they can set this value to `2px solid #f00` where we just want the width value to use in the calculation.

File: packages/terra-popup/src/PopupArrow.module.scss [hunk 4]
```diff
@@ -3,82 +3,82 @@
   [dir=rtl] {
     // These are relative to screen coordinates, so no bi-di concerns
     .popup-arrow {
-      border: var(--terra-popup-arrow-size, 11px) solid transparent;
+      border: var(--terra-popup-arrow-border, 11px) solid transparent;
```
Comment (bjankord) [inline packages/terra-popup/src/PopupArrow.module.scss]: Let's update this to:
```suggestion
      border-color: transparent;
      border-style: solid;
      border-width: var(--terra-popup-arrow-border-width, 11px); 
```

File: packages/terra-popup/src/PopupArrow.module.scss [hunk 5]
```diff
@@ -3,82 +3,82 @@
...
       left: 50%;
       position: absolute;
       top: 50%;
       width: 0;
-      z-index: var(--terra-popup-arrow-z-index, 7002);
+      z-index: 7002;
 
       &::after {
-        border: var(--terra-popup-arrow-inner-size, 10px) solid transparent;
+        border: var(--terra-popup-arrow-after-border, 10px) solid transparent;
```
Comment (bjankord) [inline packages/terra-popup/src/PopupArrow.module.scss]: Let's update this to 
```suggestion
      border-color: transparent;
      border-style: solid;
      border-width: var(--terra-popup-arrow-after-border-width, 11px); 
```

File: packages/terra-application-utility/src/utility/_UtilityMenu.module.scss [hunk 6]
```diff
@@ -19,81 +19,102 @@
...
+    word-wrap: break-word;
   }
 
   .utility-menu-icon-left {
     flex-shrink: 0;
   }
 
   .header-utility-menu-icon-left {
-    width: var(--terra-application-header-utility-menu-chevron-width, 0.7143rem);
+    width: var(--terra-application-utility-header-type-menu-back-width, 0.7143rem);
```
Comment (bjankord) [inline packages/terra-application-utility/src/utility/_UtilityMenu.module.scss]: nit: This may be more clear if we add `icon` to the name
```suggestion
    width: var(--terra-application-utility-header-type-menu-back-icon-width, 0.7143rem);
```

File: packages/terra-application-utility/src/utility/_UtilityMenu.module.scss [hunk 7]
```diff
@@ -19,81 +19,102 @@
...
   }
 
   .menu-utility-menu-icon-left {
-    width: var(--terra-application-menu-utility-menu-chevron-width, 0.7143rem);
+    width: var(--terra-application-utility-menu-type-menu-back-width, 0.7143rem);
```
Comment (bjankord) [inline packages/terra-application-utility/src/utility/_UtilityMenu.module.scss]: nit: This may be more clear if we add `icon` to the name
```suggestion
    width: var(--terra-application-utility-menu-type-menu-back-icon-width, 0.7143rem);
```

File: packages/terra-application-utility/src/utility/_UtilityMenu.module.scss [hunk 8]
```diff
@@ -19,81 +19,102 @@
...
+  }
+
   .utility-menu-icon-close {
     flex-shrink: 0;
   }
 
   .header-utility-menu-icon-close {
-    width: var(--terra-application-header-utility-menu-page-chevron-width, 1rem);
+    display: var(--terra-application-utility-header-type-menu-page-display);
+    width: var(--terra-application-utility-header-type-menu-page-close-width, 1rem);
```
Comment (bjankord) [inline packages/terra-application-utility/src/utility/_UtilityMenu.module.scss]: nit: This may be more clear if we add `icon` to the name
```suggestion
    width: var(--terra-application-utility-header-type-menu-page-close-icon-width, 1rem);
```

File: packages/terra-application-utility/src/utility/_UtilityMenu.module.scss [hunk 9]
```diff
@@ -19,81 +19,102 @@
...
   }
 
   .menu-utility-menu-icon-close {
-    width: var(--terra-application-menu-utility-menu-page-chevron-width, 1rem);
+    width: var(--terra-application-utility-menu-type-menu-page-close-width, 1rem);
```
Comment (bjankord) [inline packages/terra-application-utility/src/utility/_UtilityMenu.module.scss]: nit: This may be more clear if we add `icon` to the name
```suggestion
    width: var(--terra-application-utility-menu-type-menu-page-close-icon-width, 1rem);
```

File: packages/terra-application-header-layout/src/ApplicationHeaderLayout.module.scss [hunk 10]
```diff
@@ -16,7 +16,7 @@
     align-items: center;
     display: flex;
     height: 100%;
-    overflow: hidden;
+    overflow: visible;
```
Comment (StephenEsser) [inline packages/terra-application-header-layout/src/ApplicationHeaderLayout.module.scss]: Do you remember why this change was made?

File: packages/terra-application-links/src/tabs/ApplicationTabs.module.scss [hunk 11]
```diff
@@ -5,8 +5,9 @@
     flex-wrap: nowrap;
     height: 100%;
     justify-content: center;
-    padding-left: var(--terra-application-tabs-spacer-width, 2rem);
-    padding-right: var(--terra-application-tabs-spacer-width, 2rem);
+    overflow: hidden;
+    padding-left: var(--terra-application-links-tabs-container-padding-left, 1.4285714286rem);
+    padding-right: var(--terra-application-links-tabs-container-padding-right, 1.4285714286rem);
```
Comment (StephenEsser) [inline packages/terra-application-links/src/tabs/ApplicationTabs.module.scss]: There's a `.tabs-container` block below this block. There will be a collision if that block ever needs a left / right padding.

```scss
  .application-tabs {
    padding-left: var(--terra-application-links-tabs-container-padding-left, 1.4285714286rem);
    padding-right: var(--terra-application-links-tabs-container-padding-right, 1.4285714286rem);
  }

  .tabs-container {
    width: calc(100% - var(--terra-application-links-tabs-container-padding-left, 1.4285714286rem) - var(--terra-application-links-tabs-container-padding-right, 1.4285714286rem));
  }
```

File: packages/terra-application-links/src/tabs/ApplicationTabs.module.scss [hunk 12]
```diff
@@ -42,123 +43,180 @@
...
+      overflow: hidden;
+      padding-left: var(--terra-application-links-tabs-tab-padding-left, 1.0715rem);
+      padding-right: var(--terra-application-links-tabs-tab-padding-right, 1.0715rem);
+      position: absolute;
+      top: 0;
+      transition-duration: var(--terra-application-links-tabs-tab-transition-duration, 0s);
+      transition-property: background-size;
+      transition-timing-function: var(--terra-application-links-tabs-tab-transition-timing-function, ease);
+      width: calc(100% - var(--terra-application-links-tabs-tab-divider-before-width, 2px));
+    }
```
Comment (StephenEsser) [inline packages/terra-application-links/src/tabs/ApplicationTabs.module.scss]: This is a little tricky. This block may not scale very well in the long run. There are three different elements sharing the same name scope. 

```scss
  .tab,
  .tab-menu-display,
  .divider-after-last-tab {
    background-color: transparent; // Not themed yet but would collide with .tab-inner theme variable below
    color: var(--terra-application-links-tabs-tab-color, #fff);

    &::before {
      // This specifies divider but is applied onto .tab, .tab-menu-display, and .divider-after-last-tab
      background-color: var(--terra-application-links-tabs-tab-divider-before-background-color, #0079be);
    }

    .tab-inner {
      // This is applied to tab-inner but does not scope it to tab inner.
      background-color: var(--terra-application-links-tabs-tab-background-color, transparent);
    }
}
```

File: packages/terra-application-utility/src/utility/_UtilityMenu.module.scss (hunk 6 above)
Comment (BenBoersma) [reply packages/terra-application-utility/src/utility/_UtilityMenu.module.scss]: Fixed in https://github.com/cerner/terra-framework/pull/439/commits/efbb9e2eb8d2c426fa2993fa3bb0d20efb23c927

File: packages/terra-application-utility/src/utility/_UtilityMenu.module.scss (hunk 7 above)
Comment (BenBoersma) [reply packages/terra-application-utility/src/utility/_UtilityMenu.module.scss]: Fixed in https://github.com/cerner/terra-framework/pull/439/commits/efbb9e2eb8d2c426fa2993fa3bb0d20efb23c927

File: packages/terra-application-utility/src/utility/_UtilityMenu.module.scss (hunk 8 above)
Comment (BenBoersma) [reply packages/terra-application-utility/src/utility/_UtilityMenu.module.scss]: Fixed in https://github.com/cerner/terra-framework/pull/439/commits/efbb9e2eb8d2c426fa2993fa3bb0d20efb23c927

File: packages/terra-application-utility/src/utility/_UtilityMenu.module.scss (hunk 9 above)
Comment (BenBoersma) [reply packages/terra-application-utility/src/utility/_UtilityMenu.module.scss]: Fixed in https://github.com/cerner/terra-framework/pull/439/commits/efbb9e2eb8d2c426fa2993fa3bb0d20efb23c927

File: packages/terra-application-header-layout/src/ApplicationHeaderLayout.module.scss (hunk 10 above)
Comment (BenBoersma) [reply packages/terra-application-header-layout/src/ApplicationHeaderLayout.module.scss]: @neilpfeiffer made the change, so I am pinging him for the question.

File: packages/terra-time-input/src/TimeInput.module.scss (hunk 1 above)
Comment (BenBoersma) [reply packages/terra-time-input/src/TimeInput.module.scss]: As per offline discussion, we are leaving this as is for now. So I am gonna mark this has resolved.

File: packages/terra-application-links/src/tabs/ApplicationTabs.module.scss (hunk 11 above)
Comment (bjankord) [reply packages/terra-application-links/src/tabs/ApplicationTabs.module.scss]: Suggest updating `application-tabs` to `tabs-wrapper` so we can do something like this.

```scss
  .tabs-wrapper {
    padding-left: var(--terra-application-links-tabs-wrapper-padding-left, 1.4285714286rem);
    padding-right: var(--terra-application-links-tabs-wrapper-padding-right, 1.4285714286rem);
  }

  .tabs-container {
    width: calc(100% - var(--terra-application-links-tabs-container-padding-left, 1.4285714286rem) - var(--terra-application-links-tabs-container-padding-right, 1.4285714286rem));
  }
```

File: packages/terra-application-links/src/tabs/ApplicationTabs.module.scss (hunk 12 above)
Comment (bjankord) [reply packages/terra-application-links/src/tabs/ApplicationTabs.module.scss]: Yeah, I think from a scalability perspective, it may be better to split these selectors and variables up even though they all have the same value in this default theme.

Comment (BenBoersma) [reply packages/terra-application-links/src/tabs/ApplicationTabs.module.scss]: It looks like the `&::before` section that specifies divider on its variable names isn't intended to mean only the `.divider-after-last-tab` class but all 3 classes. It generates the borders between each tab in the tab menu.

File: packages/terra-brand-footer/src/BrandFooter.module.scss [hunk 13]
```diff
@@ -83,24 +83,24 @@
     .list-header + .menu::before {
       content: '';
       padding-left: 0;
-      padding-right: var(--terra-brand-footer-horizontal-nav-first-link-padding-before-right, 0.7142857142857143rem);
+      padding-right: var(--terra-brand-footer-horizontal-nav-first-link-before-padding-right, 0.7142857142857143rem);
     }
     /* stylelint-enable selector-max-compound-selectors */
 
     .separator {
       color: var(--terra-brand-footer-horizontal-nav-link-divider-color, #8baaca);
```
Comment (StephenEsser) [inline packages/terra-brand-footer/src/BrandFooter.module.scss]: This guy is `divider` vs `separator` below

File: packages/terra-application-links/src/tabs/ApplicationTabs.module.scss (hunk 11 above)
Comment (BenBoersma) [reply packages/terra-application-links/src/tabs/ApplicationTabs.module.scss]: Updated here https://github.com/cerner/terra-framework/pull/439/commits/ca6e18fc82385d7f2349879e6f35f25f1f502ed0

File: packages/terra-popup/src/PopupArrow.module.scss [hunk 14]
```diff
@@ -3,82 +3,86 @@
...
       transform: translate3d(-100%, -100%, 0);
 
       &::after {
         border-left-width: 0;
-        border-right-color: var(--terra-popup-arrow-inner-color, #fff);
-        left: var(--terra-popup-arrow-inner-inset-size, 2px);
-        top: calc(-1 * var(--terra-popup-arrow-inner-size, 10px));
+        border-right-color: var(--terra-popup-arrow-after-border-right-color, #fff);
+        left: var(--terra-popup-align-arrow-left-after-left, 2px);
+        top: calc(-1 * var(--terra-popup-align-arrow-left-after-top, 10px));
```
Comment (StephenEsser) [inline packages/terra-popup/src/PopupArrow.module.scss]: I'm having a hard time figuring out the intent of these. Is the additional left intentional in both of these? `left-after-left` and `left-after-top`. 

border-right-color doesn't have an `align` prefix, but the other two do.

Comment (StephenEsser) [reply packages/terra-popup/src/PopupArrow.module.scss]: Is `left` the arrow direction from the data selector above?

File: packages/terra-brand-footer/src/BrandFooter.module.scss (hunk 13 above)
Comment (BenBoersma) [reply packages/terra-brand-footer/src/BrandFooter.module.scss]: Fixed here https://github.com/cerner/terra-framework/pull/439/commits/3152c95dd5c9c39f028a92d2d2e884549d282331

File: packages/terra-popup/src/PopupArrow.module.scss (hunk 14 above)
Comment (BenBoersma) [reply packages/terra-popup/src/PopupArrow.module.scss]: So the first `left` in `left-after-left` is supposed to be arrow direction. I believe they were probably separated like this becasue the left, top, right, and bottom directions needed a secondary qualifer for them but the `border-right-color` property did not since those were unique to each arrow direction.

For clarity, we should probably maintain a unique identifier for each of them in the form of arrow direction, though it might be better to swap it to something like `--terra-popup-left-align-arrow-after-left`

Comment (bjankord) [reply packages/terra-popup/src/PopupArrow.module.scss]: Yeah, `--terra-popup-left-align-arrow-after-left` reads better in mind, easier to reason about what it is for.

File: packages/terra-application-header-layout/src/ApplicationHeaderLayout.module.scss (hunk 10 above)
Comment (bjankord) [reply packages/terra-application-header-layout/src/ApplicationHeaderLayout.module.scss]: My thought is this is to allow the box shadow to display? cc/ @neilpfeiffer

File: packages/terra-popup/src/PopupArrow.module.scss (hunk 14 above)
Comment (BenBoersma) [reply packages/terra-popup/src/PopupArrow.module.scss]: Fixed here https://github.com/cerner/terra-framework/pull/439/commits/2621fc4230b26560a600e1cc1474eb6c901de673

File: packages/terra-application-header-layout/src/ApplicationHeaderLayout.module.scss (hunk 10 above)
Comment (neilpfeiffer) [reply packages/terra-application-header-layout/src/ApplicationHeaderLayout.module.scss]: most likely that is correct.

File: packages/terra-application-links/src/tabs/ApplicationTabs.module.scss [hunk 15]
```diff
@@ -42,123 +43,180 @@
...
+
+      &::before {
+        background-color: var(--terra-application-links-tabs-tab-selected-divider-before-background-color, #0079be);
+        background-image: var(--terra-application-links-tabs-tab-selected-divider-before-background-image);
+        border-left-color: var(--terra-application-links-tabs-tab-selected-divider-before-border-left-color, rgba(255, 255, 255, 0.17));
+        border-right-color: var(--terra-application-links-tabs-tab-selected-divider-before-border-right-color, rgba(0, 0, 0, 0.17));
+      }
+
+      .tab-inner {
+        background-color: var(--terra-application-links-tabs-selected-background-color, rgba(0, 0, 0, 0.15));
```
Comment (yuderekyu) [inline packages/terra-application-links/src/tabs/ApplicationTabs.module.scss]: we should add `tab` to be consistent with the above blocks and separate `.tab-inner` block on line 119
```suggestion
        background-color: var(--terra-application-links-tabs-tab-selected-background-color, rgba(0, 0, 0, 0.15));
```

File: packages/terra-brand-footer/src/BrandFooter.module.scss [hunk 16]
```diff
@@ -80,27 +80,27 @@
 
     /* stylelint-disable selector-max-compound-selectors */
     // A '/' isn't needed to separate an item from the header it follows
-    .list-header + .menu::before {
+    .list-header + .menu::before {
```
Comment (yuderekyu) [inline packages/terra-brand-footer/src/BrandFooter.module.scss]: nit: unnecessary whitespace

File: packages/terra-slide-panel/src/SlidePanel.module.scss [hunk 17]
```diff
@@ -73,35 +74,63 @@
...
+      }
+
       &[data-slide-panel-panel-size='small'] > .main {
-        width: calc(100% - #{var(--terra-slide-panel-panel-width-small, 320px)});
+        width: calc(100% - #{var(--terra-slide-panel-panel-squish-width-small, 320px)});
       }
 
       &[data-slide-panel-panel-size='large'] > .main {
-        width: calc(100% * (1 - #{var(--terra-slide-panel-panel-width-large-scalar, 0.6)}));
+        width: calc(100% * (1 - #{var(--terra-slide-panel-panel-squish-width-large-scalar, 0.6)}));
```
Comment (yuderekyu) [inline packages/terra-slide-panel/src/SlidePanel.module.scss]: hmmm, do we need `scalar`?

Comment (bjankord) [reply packages/terra-slide-panel/src/SlidePanel.module.scss]: I'm okay with keeping scalar.

File: packages/terra-brand-footer/src/BrandFooter.module.scss (hunk 16 above)
Comment (bjankord) [reply packages/terra-brand-footer/src/BrandFooter.module.scss]: Resolved in this commit: https://github.com/cerner/terra-framework/pull/439/commits/3593894bab994b5d7bb4e09db4d380b0f8c177a6

File: packages/terra-application-links/src/tabs/ApplicationTabs.module.scss (hunk 15 above)
Comment (BenBoersma) [reply packages/terra-application-links/src/tabs/ApplicationTabs.module.scss]: Should be more consistent here
https://github.com/cerner/terra-framework/pull/439/commits/fbd5fa6578ceca17b1778f3df76dec9e33c387a4#diff-c35c5a53e93ecb7b1096d5e08a498c61R171
https://github.com/cerner/terra-framework/pull/439/commits/fbd5fa6578ceca17b1778f3df76dec9e33c387a4#diff-c35c5a53e93ecb7b1096d5e08a498c61R294

Comment (bjankord) [reply packages/terra-application-links/src/tabs/ApplicationTabs.module.scss]: Looks good to me

File: packages/terra-application-links/src/tabs/ApplicationTabs.module.scss (hunk 12 above)
Comment (BenBoersma) [reply packages/terra-application-links/src/tabs/ApplicationTabs.module.scss]: Okay as per offline discussion with @neilpfeiffer, I have undone the selector split.

**Comment by BenBoersma** [comment]
Upgrade Guides are now updated to have this PR's changes in.
